### PR TITLE
syscalls: arm: Fix possible overflow in is_in_region function

### PR DESCRIPTION
**Pull request by ceolin**
This function is widely used by functions that validate memory
buffers. Macros used to check permissions, like Z_SYSCALL_MEMORY_READ
and Z_SYSCALL_MEMORY_WRITE, use these functions to check that a
pointers passed by user threads in a syscall.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

### CODE REVIEW

File: arch/arm/core/aarch32/cortex_m/mpu/arm_mpu_v7_internal.h
Comment (stephanosio) [inline arch/arm/core/aarch32/cortex_m/mpu/arm_mpu_v7_internal.h]: `__builtin_add_overflow` may not be available on all compilers (for GCC, you need GCC 5.1 or higher).

Should we add/use a fallback macro instead like Linux kernel does?
https://github.com/torvalds/linux/blob/8b614cb8f1dcac8ca77cf4dd85f46ef3055f8238/include/linux/overflow.h#L55

Comment (stephanosio) [reply arch/arm/core/aarch32/cortex_m/mpu/arm_mpu_v7_internal.h]: Actually, it turns out we already have something similar:
https://github.com/zephyrproject-rtos/zephyr/blob/4b38659b0661962a3fb9fc98c634944d88236923/include/sys/math_extras_impl.h#L32

Comment (ceolin) [reply arch/arm/core/aarch32/cortex_m/mpu/arm_mpu_v7_internal.h]: thanks for pointing it out. Will change it.

**Comment by jhedberg** [comment]
@ceolin I've restarted Shippable a few times, but still keeps failing - is it because of something in this PR?

**Comment by jhedberg** [comment]
> @ceolin I've restarted Shippable a few times, but still keeps failing - is it because of something in this PR?

Answering myself: seems like it is:
```
INFO    - /home/buildslave/src/github.com/zephyrproject-rtos/zephyr/sanity-out/mps2_an385/tests/kernel/pipe/pipe/kernel.pipe/handler.log
ERROR   - *** Booting Zephyr OS build v2.2.0-rc3-25-g426ddd95ee73  ***
Running test suite test_pipe
===================================================================
starting test - test_pipe_on_single_elements
E: syscall z_vrfy_k_pipe_put failed check: Memory region 0x200001c1 (size 0) read access denied
E: r0/a1:  0x00000000  r1/a2:  0x00000000  r2/a3:  0x00000000
E: r3/a4:  0x00000000 r12/ip:  0x00000000 r14/lr:  0x00000000
E:  xpsr:  0x00000000
E: Faulting instruction address (r15/pc): 0x00000000
E: >>> ZEPHYR FATAL ERROR 3: Kernel oops on CPU 0
E: Current thread: 0x200005d8 (unknown)
Caught system error -- reason 3
```

**Comment by ceolin** [comment]
@johan there are two issues that we need to figure out how to handle

1) There is one pipe tests that pass a 0 data buffer. Is it valid ? 0 data buffer was causing and underflow and magically going to implementation function. I can put a new test in the verification function to return 0 in that situation and not going to check the memory buffer

2) poll test is failing for the same problem, on test with 0 events. Same thing the issue above, but here is more confuse because of the expected behavior, currently with 0 elements it returns EAGAIN when timeout expires, what is not possible to do in the verification function. I could return -EINVAL (that looks the right thing), but then the test will look really weird, because when the test executes from supervisor thread it will return -EAGAIN and when it runs from user thread will return -EINVAL.

any thoughts ?

File: arch/arm/core/aarch32/cortex_m/mpu/arm_mpu_v7_internal.h
Comment (ioannisg) [inline arch/arm/core/aarch32/cortex_m/mpu/arm_mpu_v7_internal.h]: For ease of reading, can you also add brackets around the operations in all the "if" blocks, to explicitly indicate precedence?

File: arch/arm/core/aarch32/cortex_m/mpu/arm_mpu_v7_internal.h
Comment (ioannisg) [inline arch/arm/core/aarch32/cortex_m/mpu/arm_mpu_v7_internal.h]: here, too

Comment (ioannisg) [reply arch/arm/core/aarch32/cortex_m/mpu/arm_mpu_v7_internal.h]: and i the nxp_mpu.c

**Comment by ceolin** [comment]
> @jhedberg there are two issues that we need to figure out how to handle
> 
> 1. There is one pipe tests that pass a 0 data buffer. Is it valid ? 0 data buffer was causing and underflow and magically going to implementation function. I can put a new test in the verification function to return 0 in that situation and not going to check the memory buffer
> 2. poll test is failing for the same problem, one test with 0 events. Same thing the issue above, but here is more confuse because of the expected behavior, currently with 0 elements it returns EAGAIN when timeout expires, what is not possible to do in the verification function. I could return -EINVAL (that looks the right thing), but then the test will look really weird, because when the test executes from supervisor thread it will return -EAGAIN and when it runs from user thread will return -EINVAL.
> 
> any thoughts ?

@andyross @andrewboie ideas ?

File: arch/arm/core/aarch32/cortex_m/mpu/arm_mpu_v7_internal.h
Comment (ceolin) [reply arch/arm/core/aarch32/cortex_m/mpu/arm_mpu_v7_internal.h]: can do that

Comment (ioannisg) [inline arch/arm/core/aarch32/cortex_m/mpu/arm_mpu_v7_internal.h]: Also, if you want the function to return false when size is zero, I believe this needs to be documented.
Furthermore, I think the case of zero "size", should be better done at the caller function, mpu_buffer_validate, or even documented as a non-allowed input in the common arch interface API, arch_buffer_validate.  @andrewboie what do you think?

Comment (ceolin) [reply arch/arm/core/aarch32/cortex_m/mpu/arm_mpu_v7_internal.h]: yeah, that is weird, but I believe x86 works same way, I'll change the pr to allows size 0, what means just check whether or not the buffer address is in the given region. This will fix the problem and keeps the current behavior

Comment (ceolin) [reply arch/arm/core/aarch32/cortex_m/mpu/arm_mpu_v7_internal.h]: whether or not allow size 0 can be addressed / discussed later to no block the release, since it seems not to be a bug. The questions regarding poll and pipe behavior are still open IMHO.

**Comment by jhedberg** [comment]
@ioannisg @stephanosio are you guys ok with the latest version of this PR? I'd like some more reviews than just my own before merging.

**Review by jhedberg — APPROVED** [review]

**Comment by jhedberg** [comment]
> Will check this again early tomorrow morning @jhedberg

@ioannisg any update on this?

**Review by ioannisg — COMMENTED** [review]
@ceolin correct me if I am wrong, but now, it seems that supplying a zero sized buffer will result in the function returning true.This is a change compared to the original behavior, so I wonder if we need that.
could this PR just fix the overflow issue?

**Comment by ceolin** [comment]
> @ceolin correct me if I am wrong, but now, it seems that supplying a zero sized buffer will result in the function returning true.This is a change compared to the original behavior, so I wonder if we need that.
> could this PR just fix the overflow issue?
 
We were wrongly accepting size 0.
`if (start >= r_addr_start && (start + size - 1) <= r_addr_end)`
as you can see, the second condition will always be true for size == 0

Now, at least we are not comparing and address out of the "valid" range. My initial patch was not accepting size == 0 but then it broke some tests and the behavior was different from X86. My suggestion is merge it and fix the problems and then we can decided whether or not accept 0 size buffers, this will apply to X86 and ARC as well.

**Review by tyrantlizard77 — APPROVED** [review]

**Comment by andrewboie** [comment]
Is a 0-size buffer valid inputs to these system calls? What does it mean to pass a buffer like that?

**Comment by ceolin** [comment]
> Is a 0-size buffer valid inputs to these system calls? What does it mean to pass a buffer like that?

it depends on syscall, poll for example seems to return when the timeout expires, without anything valid, pipe just return 0 

IMHO these APIs shouldn't accept 0 size buffer

**Comment by andrewboie** [comment]
> it depends on syscall, poll for example seems to return when the timeout expires, without anything valid, pipe just return 0

I understand the behavior, but not the semantics. Why is the test passing a 0 buffer in the first place? What is it testing?

**Comment by ceolin** [comment]
> > it depends on syscall, poll for example seems to return when the timeout expires, without anything valid, pipe just return 0
> 
> I understand the behavior, but not the semantics. Why is the test passing a 0 buffer in the first place? What is it testing?

The test is testing this corner case what is valid. It is a boundary that has to be tested, the problem is if we reject 0 size buffer at memory access level we trigger an exception what the tests and APIs don't expect

**Comment by andrewboie** [comment]
> The test is testing this corner case what is valid. It is a boundary that has to be tested, the problem is if we reject 0 size buffer at memory access level we trigger an exception what the tests and APIs don't expect

So is a 0 size buffer valid or not?

**Comment by andrewboie** [comment]
And what does it mean to call these APIs with a 0 size buffer?

**Comment by ceolin** [comment]
> > The test is testing this corner case what is valid. It is a boundary that has to be tested, the problem is if we reject 0 size buffer at memory access level we trigger an exception what the tests and APIs don't expect
> 
> So is a 0 size buffer valid or not?

That is the trick part, currently it is, but not explicitly.  Regarding the second question, the meaning is   per basis API. Take a look in `k_pipe_put`

```
/**
 * @brief Write data to a pipe.
 *
 * This routine writes up to @a bytes_to_write bytes of data to @a pipe.
 *
 * @param pipe Address of the pipe.
 * @param data Address of data to write.
 * @param bytes_to_write Size of data (in bytes).
 * @param bytes_written Address of area to hold the number of bytes written.
 * @param min_xfer Minimum number of bytes to write.
 * @param timeout Non-negative waiting period to wait for the data to be written
 *                (in milliseconds), or one of the special values K_NO_WAIT
 *                and K_FOREVER.
 *
 * @retval 0 At least @a min_xfer bytes of data were written.
 * @retval -EIO Returned without waiting; zero data bytes were written.
 * @retval -EAGAIN Waiting period timed out; between zero and @a min_xfer
 *                 minus one data bytes were written.
 * @req K-PIPE-002
 */

According with it, if `bytes_to_write` and `min_xfer` are 0 this function should return 0, what is valid. The test is testing it. This one is easy because we can't test for this case before testing the memory access, but poll is more complicate, because it accepts 0 events and just return when the timeout is reached.

**Comment by andrewboie** [comment]
I see. Maybe for footprint/performance reasons we should leave (and document in the header) that arch_buffer_validate() on a 0-sized region is undefined behavior, and for these particular syscalls add explicit checks. What do you think? We can contemplate this for another PR. Regardless of what we decide, the API docs should be updated.

**Comment by ceolin** [comment]
> I see. Maybe for footprint/performance reasons we should leave (and document in the header) that arch_buffer_validate() on a 0-sized region is undefined behavior, and for these particular syscalls add explicit checks. What do you think? We can contemplate this for another PR. Regardless of what we decide, the API docs should be updated.

Agree. Further we can have the initial patch that invalidates 0 size buffers and properly fix APIs. 

Do you want me to update this commit with that info ?

**Comment by ioannisg** [comment]
@andrewboie @ceolin I would like us to update the API and agree on the behavior when the passed size is zero before merging this PR, since this is an essential part of the patch. Ideally, I would like the test-cases and the doc-fix to come within this PR :)

**Review by ioannisg — COMMENTED** [review]
Please, submit also the fix for API documentation, as well as a test that covers this corner case of zero size.

**Comment by ceolin** [comment]
> Please, submit also the fix for API documentation, as well as a test that covers this corner case of zero size.

doing it.

**Comment by ceolin** [comment]
> Please, submit also the fix for API documentation, as well as a test that covers this corner case of zero size.

what exactly you want here ? I'll document that 0 size buffer is undefined behavior, we should not have test for it. That's said, pipe and poll tests are exercising this case, my opinion is that we fix  these APIs (maybe others) later.

**Comment by andrewboie** [comment]
It's not like this UB represents a security concern unless I'm missing something. Should not block this PR.

File: include/sys/arch_interface.h
Comment (ioannisg) [inline include/sys/arch_interface.h]: That's exactly what I wanted :)

**Review by ioannisg — APPROVED** [review]

**Comment by ioannisg** [comment]
>That's said, pipe and poll tests are exercising this case, my opinion is that we fix these APIs (maybe others) later.

Since we are now documenting that arch_buffer_validate cannot get a 0 size buffer, we need to open an issue for fixing the APIs related to the tests mentioned above

**Pull request merged**
